File: ios/StillMoment/Domain/Models/GongSound.swift
//
//  GongSound.swift
//  Still Moment
//
//  Domain - Gong Sound Model
//

import Foundation

/// Represents a configurable gong sound for meditation timer
///
/// This model defines a gong sound that can be selected for start/end gong
/// or interval gong. Sounds include:
/// - Unique identifier for persistence
/// - Audio file reference
/// - Localized name (resolved via NSLocalizedString)
///
/// Example usage:
/// ```swift
/// let sound = GongSound.defaultSound
/// print(sound.name) // "Temple Bell" (en) / "Tempelglocke" (de)
/// ```
struct GongSound: Identifiable, Equatable {
    /// Unique identifier for the sound
    let id: String

    /// Filename in the GongSounds bundle (e.g., "singing-bowl-hit-3-33366-5s.mp3")
    let filename: String

    /// Localized display name (resolved at creation time)
    let name: String
}

// MARK: - Static Sound Definitions

extension GongSound {
    /// All available gong sounds (for start/end gong)
    static let allSounds: [GongSound] = [
        GongSound(
            id: "temple-bell",
            filename: "tibetan-singing-bowl-55786-10s.mp3",
            name: NSLocalizedString("gong.temple-bell", comment: "")
        ),
        GongSound(
            id: "classic-bowl",
            filename: "singing-bowl-hit-3-33366-10s.mp3",
            name: NSLocalizedString("gong.classic-bowl", comment: "")
        ),
        GongSound(
            id: "deep-resonance",
            filename: "singing-bowl-male-frequency-29714-10s.mp3",
            name: NSLocalizedString("gong.deep-resonance", comment: "")
        ),
        GongSound(
            id: "clear-strike",
            filename: "singing-bowl-strike-sound-84682-10s.mp3",
            name: NSLocalizedString("gong.clear-strike", comment: "")
        )
    ]

    /// Soft interval tone (5th sound option, only for interval gongs)
    static let softIntervalTone = GongSound(
        id: "soft-interval",
        filename: "interval.mp3",
        name: NSLocalizedString("gong.soft-interval", comment: "")
    )

    /// All available interval gong sounds (4 standard + soft interval tone)
    static let allIntervalSounds: [GongSound] = allSounds + [softIntervalTone]

    /// Default gong sound (Temple Bell)
    static let defaultSound: GongSound = allSounds.first { $0.id == defaultSoundId } ?? allSounds[0]

    /// Default sound ID for settings
    static let defaultSoundId: String = "temple-bell"

    /// Default interval sound ID (soft interval tone)
    static let defaultIntervalSoundId: String = "soft-interval"

    /// Find a gong sound by ID (searches all sounds including interval sounds)
    /// - Parameter id: The sound ID to search for
    /// - Returns: The matching GongSound or nil if not found
    static func find(byId id: String) -> GongSound? {
        self.allIntervalSounds.first { $0.id == id }
    }

    /// Find a gong sound by ID, returning default if not found
    /// - Parameter id: The sound ID to search for
    /// - Returns: The matching GongSound or defaultSound if not found
    static func findOrDefault(byId id: String) -> GongSound {
        self.find(byId: id) ?? self.defaultSound
    }
}

File: android/app/src/main/kotlin/com/stillmoment/domain/models/GongSound.kt
package com.stillmoment.domain.models

/**
 * Represents a configurable gong sound for meditation timer.
 *
 * This model defines a gong sound that can be selected for start/end gong
 * or interval gong. Sounds include:
 * - Unique identifier for persistence
 * - Audio resource reference
 * - Localized name (German + English)
 *
 * @property id Unique identifier for the sound (used for persistence)
 * @property nameEnglish English display name
 * @property nameGerman German display name
 * @property rawResourceName Raw resource name for the audio file (e.g., "gong_temple_bell")
 */
data class GongSound(
    val id: String,
    val nameEnglish: String,
    val nameGerman: String,
    val rawResourceName: String
) {
    companion object {
        /** Default gong sound ID */
        const val DEFAULT_SOUND_ID = "temple-bell"

        /** ID for the soft interval tone (uses existing interval.mp3) */
        const val SOFT_INTERVAL_SOUND_ID = "soft-interval"

        /** ID for vibration signal (no audio — device vibration) */
        const val VIBRATION_ID = "vibration"

        private val vibrationSound = GongSound(
            id = VIBRATION_ID,
            nameEnglish = "Vibration",
            nameGerman = "Vibration",
            rawResourceName = ""
        )

        /** All available gong sounds (for start/end gong selection) */
        val allSounds: List<GongSound> = listOf(
            GongSound(
                id = "temple-bell",
                nameEnglish = "Temple Bell",
                nameGerman = "Tempelglocke",
                rawResourceName = "gong_temple_bell"
            ),
            GongSound(
                id = "classic-bowl",
                nameEnglish = "Classic Bowl",
                nameGerman = "Klassisch",
                rawResourceName = "gong_classic_bowl"
            ),
            GongSound(
                id = "deep-resonance",
                nameEnglish = "Deep Resonance",
                nameGerman = "Tiefe Resonanz",
                rawResourceName = "gong_deep_resonance"
            ),
            GongSound(
                id = "clear-strike",
                nameEnglish = "Clear Strike",
                nameGerman = "Klarer Anschlag",
                rawResourceName = "gong_clear_strike"
            ),
            vibrationSound
        )

        /** All available interval sounds (soft interval tone first, then allSounds, vibration last) */
        val allIntervalSounds: List<GongSound> = listOf(
            GongSound(
                id = SOFT_INTERVAL_SOUND_ID,
                nameEnglish = "Soft Interval Tone",
                nameGerman = "Sanfter Intervallton",
                rawResourceName = "interval"
            ),
            GongSound(
                id = "temple-bell",
                nameEnglish = "Temple Bell",
                nameGerman = "Tempelglocke",
                rawResourceName = "gong_temple_bell"
            ),
            GongSound(
                id = "classic-bowl",
                nameEnglish = "Classic Bowl",
                nameGerman = "Klassisch",
                rawResourceName = "gong_classic_bowl"
            ),
            GongSound(
                id = "deep-resonance",
                nameEnglish = "Deep Resonance",
                nameGerman = "Tiefe Resonanz",
                rawResourceName = "gong_deep_resonance"
            ),
            GongSound(
                id = "clear-strike",
                nameEnglish = "Clear Strike",
                nameGerman = "Klarer Anschlag",
                rawResourceName = "gong_clear_strike"
            ),
            vibrationSound
        )

        /** Default gong sound (Temple Bell) */
        val defaultSound: GongSound = allSounds.first { it.id == DEFAULT_SOUND_ID }

        /**
         * Find a gong sound by ID (searches all sounds including interval-only sounds).
         *
         * @param id The sound ID to search for
         * @return The matching GongSound or null if not found
         */
        fun find(id: String): GongSound? = allIntervalSounds.find { it.id == id }

        /**
         * Find a gong sound by ID, returning default if not found.
         *
         * @param id The sound ID to search for
         * @return The matching GongSound or defaultSound if not found
         */
        fun findOrDefault(id: String): GongSound = find(id) ?: defaultSound
    }
}
